core/client/src/client.rs

```diff
@@ -1069,8 +1080,17 @@ impl<B, E, Block, RA> Client<B, E, Block, RA> where
 		// if the block is not a direct ancestor of the current best chain,
 		// then some other block is the common ancestor.
 		if route_from_best.common_block().hash != block {
-			// FIXME: #1442 reorganize best block to be the best chain containing
-			// `block`.
+			let new_best = match select_chain {
+				// find the new best block that is a descendent of the finalized
+				// block. we already have the import lock acquired so there's no
+				// need to re-acquire it in `select_chain`.
+				Some(select_chain) => select_chain.best_containing(block, None, None)?
+					.unwrap_or(block),
+				// if no `select_chain` is provided set the finalized block as best
+				None => block,
```

Comment:
In some situations, this is clearly OK, but I would worry that `None` is passed even in situations where we need it. Rather than use an `Option`, I recommend creating a dummy `SelectChain` for the cases where we currently return `None`, which always returns the known best block for `best_containing`.